Support for compiling with clang++ >= 3.9.1

We currently hit errors due to warnings with clang++ 3.9.1. This is because the `enum.hpp` header emits a warning with `-Wundefined-var-template` which is newly added to clang++ and enabled by default: http://llvm.org/viewvc/llvm-project?view=revision&revision=266719.

@kkaefer @jfirebaugh - any ideas to work around? Or should we suppress?

```
../../../include/mbgl/util/enum.hpp:17:32: error: instantiation of variable 'mbgl::Enum<mbgl::style::SymbolPlacementType>::begin' required here, but no definition is available [-Werror,-Wundefined-var-template]
        auto it = std::find_if(begin, end, [&] (const auto& v) { return t == v.first; });
                               ^
../../../src/mbgl/style/conversion/stringify.hpp:52:28: note: in instantiation of member function 'mbgl::Enum<mbgl::style::SymbolPlacementType>::toString' requested here
    writer.String(Enum<T>::toString(v));
                           ^
../../../src/mbgl/style/conversion/stringify.hpp:236:39: note: in instantiation of function template specialization 'mbgl::style::conversion::stringify<rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::CrtAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator, 0>, mbgl::style::SymbolPlacementType, void>' requested here
    v.evaluate([&] (const auto& v_) { stringify(writer, v_); });
                                      ^
../../../mason_packages/headers/variant/1.1.4/include/mapbox/variant.hpp:304:51: note: in instantiation of member function 'mapbox::util::detail::dispatcher<const (lambda at ../../../src/mbgl/style/conversion/stringify.hpp:236:16) &, mapbox::util::variant<mbgl::style::Undefined, mbgl::style::SymbolPlacementType, mbgl::style::Function<mbgl::style::SymbolPlacementType> >, void, mbgl::style::SymbolPlacementType, mbgl::style::Function<mbgl::style::SymbolPlacementType> >::apply_const' requested here
            return dispatcher<F, V, R, Types...>::apply_const(v, std::forward<F>(f));
                                                  ^
../../../mason_packages/headers/variant/1.1.4/include/mapbox/variant.hpp:850:55: note: in instantiation of member function 'mapbox::util::detail::dispatcher<const (lambda at ../../../src/mbgl/style/conversion/stringify.hpp:236:16) &, mapbox::util::variant<mbgl::style::Undefined, mbgl::style::SymbolPlacementType, mbgl::style::Function<mbgl::style::SymbolPlacementType> >, void, mbgl::style::Undefined, mbgl::style::SymbolPlacementType, mbgl::style::Function<mbgl::style::SymbolPlacementType> >::apply_const' requested here
        return detail::dispatcher<F, V, R, Types...>::apply_const(v, std::forward<F>(f));
                                                      ^
../../../include/mbgl/style/property_value.hpp:44:23: note: in instantiation of function template specialization 'mapbox::util::variant<mbgl::style::Undefined, mbgl::style::SymbolPlacementType, mbgl::style::Function<mbgl::style::SymbolPlacementType> >::visit<const (lambda at ../../../src/mbgl/style/conversion/stringify.hpp:236:16) &, mapbox::util::variant<mbgl::style::Undefined, mbgl::style::SymbolPlacementType, mbgl::style::Function<mbgl::style::SymbolPlacementType> >, void>' requested here
        return Value::visit(value, evaluator);
                      ^
../../../src/mbgl/style/conversion/stringify.hpp:243:9: note: in instantiation of function template specialization 'mbgl::style::conversion::stringify<rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::CrtAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator, 0>, mbgl::style::SymbolPlacementType>' requested here
        stringify(writer, value);
        ^
../../../src/mbgl/style/conversion/stringify.hpp:250:21: note: in instantiation of function template specialization 'mbgl::style::conversion::stringify<mbgl::style::SymbolPlacement, rapidjson::Writer<rapidjson::GenericStringBuffer<rapidjson::UTF8<char>, rapidjson::CrtAllocator>, rapidjson::UTF8<char>, rapidjson::UTF8<char>, rapidjson::CrtAllocator, 0>, mbgl::style::SymbolPlacementType>' requested here
    util::ignore({ (stringify<Ps>(writer, ps.unevaluated.template get<Ps>()), 0)... });
                    ^
```

A second issue needing addressed to support clang >= 3.9.1 is this failure in the tests:

```
/home/travis/build/mapbox/mapbox-gl-native/test/api/annotations.test.cpp:76:32: error: no matching constructor for initialization of 'const mapbox::geometry::linear_ring<double, std::vector>'
    Polygon<double> polygon = {{ {{ { 0, 0 }, { 0, 45 }, { 45, 45 }, { 45, 0 } }} }};
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/usr/lib/gcc/x86_64-linux-gnu/5.4.1/../../../../include/c++/5.4.1/bits/stl_vector.h:264:7: note: candidate inherited constructor not viable: cannot convert initializer list argument to 'const allocator_type' (aka 'const std::allocator<mapbox::geometry::point<double> >')
      vector(const allocator_type& __a) noexcept
      ^
../../../mason_packages/headers/geometry/0.9.0/include/mapbox/geometry/polygon.hpp:18:27: note: constructor from base class 'vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >' inherited here
    using container_type::container_type;
                          ^
/usr/lib/gcc/x86_64-linux-gnu/5.4.1/../../../../include/c++/5.4.1/bits/stl_vector.h:277:7: note: candidate inherited constructor not viable: cannot convert initializer list argument to 'size_type' (aka 'unsigned long')
      vector(size_type __n, const allocator_type& __a = allocator_type())
      ^
../../../mason_packages/headers/geometry/0.9.0/include/mapbox/geometry/polygon.hpp:18:27: note: constructor from base class 'vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >' inherited here
    using container_type::container_type;
                          ^
/usr/lib/gcc/x86_64-linux-gnu/5.4.1/../../../../include/c++/5.4.1/bits/stl_vector.h:318:7: note: candidate inherited constructor not viable: cannot convert initializer list argument to 'const std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >'
      vector(const vector& __x)
      ^
../../../mason_packages/headers/geometry/0.9.0/include/mapbox/geometry/polygon.hpp:18:27: note: constructor from base class 'vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >' inherited here
    using container_type::container_type;
                          ^
/usr/lib/gcc/x86_64-linux-gnu/5.4.1/../../../../include/c++/5.4.1/bits/stl_vector.h:335:7: note: candidate inherited constructor not viable: cannot convert initializer list argument to 'std::vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >'
      vector(vector&& __x) noexcept
      ^
../../../mason_packages/headers/geometry/0.9.0/include/mapbox/geometry/polygon.hpp:18:27: note: constructor from base class 'vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >' inherited here
    using container_type::container_type;
                          ^
/usr/lib/gcc/x86_64-linux-gnu/5.4.1/../../../../include/c++/5.4.1/bits/stl_vector.h:373:7: note: candidate inherited constructor not viable: cannot convert initializer list argument to 'mapbox::geometry::point<double>'
      vector(initializer_list<value_type> __l,
      ^
../../../mason_packages/headers/geometry/0.9.0/include/mapbox/geometry/polygon.hpp:18:27: note: constructor from base class 'vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >' inherited here
    using container_type::container_type;
                          ^
../../../mason_packages/headers/geometry/0.9.0/include/mapbox/geometry/polygon.hpp:13:8: note: candidate constructor (the implicit copy constructor) not viable: cannot convert initializer list argument to 'const mapbox::geometry::linear_ring<double, std::vector>'
struct linear_ring : Cont<point<T>>
       ^
../../../mason_packages/headers/geometry/0.9.0/include/mapbox/geometry/polygon.hpp:13:8: note: candidate constructor (the implicit move constructor) not viable: cannot convert initializer list argument to 'mapbox::geometry::linear_ring<double, std::vector>'
struct linear_ring : Cont<point<T>>
       ^
/usr/lib/gcc/x86_64-linux-gnu/5.4.1/../../../../include/c++/5.4.1/bits/stl_vector.h:401:9: note: candidate inherited constructor template not viable: requires at least 2 arguments, but 1 was provided
        vector(_InputIterator __first, _InputIterator __last,
        ^
../../../mason_packages/headers/geometry/0.9.0/include/mapbox/geometry/polygon.hpp:18:27: note: constructor from base class 'vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >' inherited here
    using container_type::container_type;
                          ^
../../../mason_packages/headers/geometry/0.9.0/include/mapbox/geometry/polygon.hpp:13:8: note: candidate constructor (the implicit default constructor) not viable: requires 0 arguments, but 1 was provided
struct linear_ring : Cont<point<T>>
       ^
/usr/lib/gcc/x86_64-linux-gnu/5.4.1/../../../../include/c++/5.4.1/bits/stl_vector.h:339:7: note: candidate inherited constructor not viable: requires 2 arguments, but 1 was provided
      vector(const vector& __x, const allocator_type& __a)
      ^
../../../mason_packages/headers/geometry/0.9.0/include/mapbox/geometry/polygon.hpp:18:27: note: constructor from base class 'vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >' inherited here
    using container_type::container_type;
                          ^
/usr/lib/gcc/x86_64-linux-gnu/5.4.1/../../../../include/c++/5.4.1/bits/stl_vector.h:348:7: note: candidate inherited constructor not viable: requires 2 arguments, but 1 was provided
      vector(vector&& __rv, const allocator_type& __m)
      ^
../../../mason_packages/headers/geometry/0.9.0/include/mapbox/geometry/polygon.hpp:18:27: note: constructor from base class 'vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >' inherited here
    using container_type::container_type;
                          ^
/usr/lib/gcc/x86_64-linux-gnu/5.4.1/../../../../include/c++/5.4.1/bits/stl_vector.h:289:7: note: candidate inherited constructor not viable: requires at least 2 arguments, but 1 was provided
      vector(size_type __n, const value_type& __value,
      ^
../../../mason_packages/headers/geometry/0.9.0/include/mapbox/geometry/polygon.hpp:18:27: note: constructor from base class 'vector<mapbox::geometry::point<double>, std::allocator<mapbox::geometry::point<double> > >' inherited here
    using container_type::container_type;
                          ^
/usr/lib/gcc/x86_64-linux-gnu/5.4.1/../../../../include/c++/5.4.1/bits/stl_vector.h:373:43: note: passing argument to parameter '__l' here
      vector(initializer_list<value_type> __l,
                                          ^
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -5,20 +5,20 @@ git:
 addons:
   apt:
     sources:
-      - &common_sources [ 'ubuntu-toolchain-r-test', 'george-edison55-precise-backports' ]
+      - &common_sources [ 'ubuntu-toolchain-r-test', 'george-edison55-precise-backports', 'llvm-toolchain-trusty-3.9' ]
     packages:
       - &common_packages [ 'libllvm3.8v4', 'cmake', 'cmake-data' ]
-      - &clang38_packages [ 'clang-3.8', 'libstdc++-5-dev', 'libstdc++6' ]
+      - &clang39_packages [ 'clang-3.9', 'libstdc++-5-dev', 'libstdc++6' ]
       - &gcc5_packages [ 'gcc-5', 'g++-5' ]
       - &glfw_packages [ 'libxrandr-dev', 'libxcursor-dev', 'libxinerama-dev' ]
 
 addons_shortcuts:
-  addons_clang38: &clang38
+  addons_clang39: &clang39
     apt:
       sources: *common_sources
       packages:
         - *common_packages
-        - *clang38_packages
+        - *clang39_packages
         - *glfw_packages
   addons_gcc5: &gcc5
     apt:
@@ -81,14 +81,14 @@ matrix:
         - git fetch origin master:refs/remotes/origin/master
         - make check
 
-    # EGL - Node - Clang 3.8 - Debug
+    # EGL - Node - Clang 3.9 - Debug
     - os: linux
       sudo: required
       dist: trusty
       language: node
-      compiler: "egl-node4-clang38-debug"
-      env: BUILDTYPE=Debug _CXX=clang++-3.8 _CC=clang-3.8 WITH_EGL=1
-      addons: *clang38
+      compiler: "egl-node4-clang39-debug"
+      env: BUILDTYPE=Debug _CXX=clang++-3.9 _CC=clang-3.9 WITH_EGL=1
+      addons: *clang39
       before_script:
         - mapbox_install_logbt
         - mapbox_start_xvfb
@@ -104,14 +104,14 @@ matrix:
       after_failure:
         - aws s3 cp . s3://mapbox/mapbox-gl-native/render-tests/$TRAVIS_JOB_NUMBER --recursive --exclude "*" --include "*.trace"
 
-    # EGL - Node - Clang 3.8 - Release
+    # EGL - Node - Clang 3.9 - Release
     - os: linux
       sudo: required
       dist: trusty
       language: node
-      compiler: "egl-node4-clang38-release"
-      env: BUILDTYPE=Release _CXX=clang++-3.8 _CC=clang-3.8 WITH_EGL=1
-      addons: *clang38
+      compiler: "egl-node4-clang39-release"
+      env: BUILDTYPE=Release _CXX=clang++-3.9 _CC=clang-3.9 WITH_EGL=1
+      addons: *clang39
       before_script:
         # fglrx causes the GLX extension to be unavailable
         - sudo apt-get purge -qq fglrx
@@ -148,14 +148,14 @@ matrix:
         - ccache --show-stats
         - ./platform/linux/scripts/coveralls.sh
 
-    # EGL - Clang 3.8 - Debug
+    # EGL - Clang 3.9 - Debug
     - os: linux
       sudo: required
       dist: trusty
       language: cpp
-      compiler: "egl-clang38-debug"
-      env: BUILDTYPE=Debug _CXX=clang++-3.8 _CC=clang-3.8 WITH_EGL=1
-      addons: *clang38
+      compiler: "egl-clang39-debug"
+      env: BUILDTYPE=Debug _CXX=clang++-3.9 _CC=clang-3.9 WITH_EGL=1
+      addons: *clang39
       before_script:
         - mapbox_start_xvfb
         - mapbox_export_mesa_library_path

--- a/include/mbgl/util/enum.hpp
+++ b/include/mbgl/util/enum.hpp
@@ -11,26 +11,26 @@ namespace mbgl {
 template <typename T>
 class Enum {
 public:
-    using Value = std::pair<const T, const char *>;
-
-    static const char * toString(T t) {
-        auto it = std::find_if(begin, end, [&] (const auto& v) { return t == v.first; });
-        assert(it != end); return it->second;
-    }
-
-    static optional<T> toEnum(const std::string& s) {
-        auto it = std::find_if(begin, end, [&] (const auto& v) { return s == v.second; });
-        return it == end ? optional<T>() : it->first;
-    }
-
-private:
-    static const Value* begin;
-    static const Value* end;
+    static const char * toString(T);
+    static optional<T> toEnum(const std::string&);
 };
 
-#define MBGL_DEFINE_ENUM(type, strings...) \
-const constexpr Enum<type>::Value type##_names[] = strings; \
-template <> const Enum<type>::Value* Enum<type>::begin = std::begin(type##_names); \
-template <> const Enum<type>::Value* Enum<type>::end = std::end(type##_names)
+#define MBGL_DEFINE_ENUM(T, values...)                                          \
+                                                                                \
+static const constexpr std::pair<const T, const char *> T##_names[] = values;   \
+                                                                                \
+template <>                                                                     \
+const char * Enum<T>::toString(T t) {                                           \
+    auto it = std::find_if(std::begin(T##_names), std::end(T##_names),          \
+        [&] (const auto& v) { return t == v.first; });                          \
+    assert(it != std::end(T##_names)); return it->second;                       \
+}                                                                               \
+                                                                                \
+template <>                                                                     \
+optional<T> Enum<T>::toEnum(const std::string& s) {                             \
+    auto it = std::find_if(std::begin(T##_names), std::end(T##_names),          \
+        [&] (const auto& v) { return s == v.second; });                         \
+    return it == std::end(T##_names) ? optional<T>() : it->first;               \
+}
 
 } // namespace mbgl

--- a/test/api/annotations.test.cpp
+++ b/test/api/annotations.test.cpp
@@ -73,7 +73,7 @@ TEST(Annotations, LineAnnotation) {
 TEST(Annotations, FillAnnotation) {
     AnnotationTest test;
 
-    Polygon<double> polygon = {{ {{ { 0, 0 }, { 0, 45 }, { 45, 45 }, { 45, 0 } }} }};
+    Polygon<double> polygon = { {{ { 0, 0 }, { 0, 45 }, { 45, 45 }, { 45, 0 } }} };
     FillAnnotation annotation { polygon };
     annotation.color = Color::red();
 
@@ -98,7 +98,7 @@ TEST(Annotations, AntimeridianAnnotationSmall) {
     lineAnnotation.width = { 2 };
     test.map.addAnnotation(lineAnnotation);
 
-    Polygon<double> polygon = {{ {{ { antimeridian+10, 0 }, { antimeridian - 10, 10 }, { antimeridian-10, -10 } }} }};
+    Polygon<double> polygon = { {{ { antimeridian+10, 0 }, { antimeridian - 10, 10 }, { antimeridian-10, -10 } }} };
     FillAnnotation polygonAnnotation { polygon };
     polygonAnnotation.color = Color::blue();
     test.map.addAnnotation(polygonAnnotation);
@@ -119,7 +119,7 @@ TEST(Annotations, AntimeridianAnnotationLarge) {
     lineAnnotation.width = { 2 };
     test.map.addAnnotation(lineAnnotation);
 
-    Polygon<double> polygon = {{ {{ { antimeridian-10, 0 }, { -antimeridian+10, 10 }, { -antimeridian+10, -10 } }} }};
+    Polygon<double> polygon = { {{ { antimeridian-10, 0 }, { -antimeridian+10, 10 }, { -antimeridian+10, -10 } }} };
     FillAnnotation polygonAnnotation { polygon };
     polygonAnnotation.color = Color::blue();
     test.map.addAnnotation(polygonAnnotation);
@@ -130,7 +130,7 @@ TEST(Annotations, AntimeridianAnnotationLarge) {
 TEST(Annotations, OverlappingFillAnnotation) {
     AnnotationTest test;
 
-    Polygon<double> polygon = {{ {{ { 0, 0 }, { 0, 45 }, { 45, 45 }, { 45, 0 } }} }};
+    Polygon<double> polygon = { {{ { 0, 0 }, { 0, 45 }, { 45, 45 }, { 45, 0 } }} };
     FillAnnotation underlaidAnnotation { polygon };
     underlaidAnnotation.color = Color::green();
     FillAnnotation overlaidAnnotation { polygon };
@@ -145,7 +145,7 @@ TEST(Annotations, OverlappingFillAnnotation) {
 TEST(Annotations, StyleSourcedShapeAnnotation) {
     AnnotationTest test;
 
-    Polygon<double> polygon = {{ {{ { 0, 0 }, { 0, 45 }, { 45, 45 }, { 45, 0 } }} }};
+    Polygon<double> polygon = { {{ { 0, 0 }, { 0, 45 }, { 45, 45 }, { 45, 0 } }} };
 
     test.map.setStyleJSON(util::read_file("test/fixtures/api/annotation.json"));
     test.map.addAnnotation(StyleSourcedAnnotation { polygon, "annotation" });
@@ -171,7 +171,7 @@ TEST(Annotations, NonImmediateAdd) {
     test.map.setStyleJSON(util::read_file("test/fixtures/api/empty.json"));
     test::render(test.map, test.view);
 
-    Polygon<double> polygon = {{ {{ { 0, 0 }, { 0, 45 }, { 45, 45 }, { 45, 0 } }} }};
+    Polygon<double> polygon = { {{ { 0, 0 }, { 0, 45 }, { 45, 45 }, { 45, 0 } }} };
     FillAnnotation annotation { polygon };
     annotation.color = Color::red();
 
@@ -245,7 +245,7 @@ TEST(Annotations, UpdateLineAnnotationStyle) {
 TEST(Annotations, UpdateFillAnnotationGeometry) {
     AnnotationTest test;
 
-    FillAnnotation annotation { Polygon<double> {{ {{ { 0, 0 }, { 0, 45 }, { 45, 45 }, { 45, 0 } }} }} };
+    FillAnnotation annotation { Polygon<double> { {{ { 0, 0 }, { 0, 45 }, { 45, 45 }, { 45, 0 } }} } };
     annotation.color = Color::red();
 
     test.map.setStyleJSON(util::read_file("test/fixtures/api/empty.json"));
@@ -253,7 +253,7 @@ TEST(Annotations, UpdateFillAnnotationGeometry) {
 
     test::render(test.map, test.view);
 
-    annotation.geometry = Polygon<double> {{ {{ { 0, 0 }, { 0, 45 }, { 45, 0 } }} }};
+    annotation.geometry = Polygon<double> { {{ { 0, 0 }, { 0, 45 }, { 45, 0 } }} };
     test.map.updateAnnotation(fill, annotation);
     test.checkRendering("update_fill_geometry");
 }
@@ -261,7 +261,7 @@ TEST(Annotations, UpdateFillAnnotationGeometry) {
 TEST(Annotations, UpdateFillAnnotationStyle) {
     AnnotationTest test;
 
-    Polygon<double> polygon = {{ {{ { 0, 0 }, { 0, 45 }, { 45, 45 }, { 45, 0 } }} }};
+    Polygon<double> polygon = { {{ { 0, 0 }, { 0, 45 }, { 45, 45 }, { 45, 0 } }} };
     FillAnnotation annotation { polygon };
     annotation.color = Color::red();
 

